### PR TITLE
Add CONTRIBUTING.md file

### PR DESCRIPTION
**Pull request by piponazo**
As we discussed in #152 , in this PR I am adding a CONTRIBUTING.md file.

Please, be aware of my limited English skills 📜

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# How to contribute #
+
+All help is appreciated. All contributors are unpaid volunteers. If you really want to see something added to Exiv2 - volunteer!
+By participating in this project, you agree to abide by the thoughtbot [code of conduct.](https://thoughtbot.com/open-source-code-of-conduct)
```
Comment (D4N) [inline CONTRIBUTING.md]: Do we really need this? imho this can be summarized by "don't be an asshole and don't discriminate". And I wouldn't link to an external document that we have no control of, since thoughbot could change that document without notice to our disadvantage.

**Comment by D4N** [comment]
This is very good! Thanks for your work!

I would propose that we require some form of tests for new code. Otherwise we'll never get the test coverage up.

I'd like to add the following coding guidelines:
- All new code that is added must be resistant to integer overflows, thus if you multiply, add, subtract, divide or bitshift integers you must ensure that no overflow can occur. Please keep in mind that signed integer overflow is **undefined behavior**, thus you must check for overflows **before** performing the arithmetic operation, otherwise the compiler is free to optimize your check after the overflow away (this has happened already). (I'll add helper functions for this and will update this file)
- All new code must be resistant to buffer overflows. Thus before you access arrays a range check must be performed.
- Distrust any data that you extract from images or from external sources. E.g. if the metadata of an image gives you an offset of another information inside that file, do not assume that this offset will not result in an out off bounds read.
- New code must not assume the endianes and the wordsize of the system it is being run on. I.e. don't assume that `sizeof(int) = 8` or that the following will work:
```
const uint32_t some_var = get_var();
const uint16_t lower_2_bytes = (const uint16_t*) &some_var;
```
since this will give you the upper two bytes on big endian systems.
If in doubt, use the fixed size integer types like `int32_t`, `uint64_t`, `size_t`, etc.
- Do not invoke undefined behavior! Ensure that with UBSAN, i.e. compile your code with `-fsanitize=undefined` and run the test suite.
- Ensure that your code has no memory errors. Use ASAN for that, i.e. compile your code with `-fsanitize=address`

That's all that I can currently think of, but I have probably forgotten a something.

File: CONTRIBUTING.md (hunk 1 above)
Comment (piponazo) [reply CONTRIBUTING.md]: Right. I just saw this in some of the examples that github recommends to read when writing the CONTRIBUTING document and I thought it was interesting. I will remove that line to be more concise.

**Comment by piponazo** [comment]
Thanks for your help :). I just added your comments + another comment about the "obligation" of testing the new code.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+- All new code that is added must be resistant to integer overflows, thus if you multiply, add, subtract, divide or bitshift integers you must ensure that no overflow can occur. Please keep in mind that signed integer overflow is undefined behavior, thus you must check for overflows before performing the arithmetic operation, otherwise the compiler is free to optimize your check after the overflow away (this has happened already).
+- All new code must be resistant to buffer overflows. Thus before you access arrays a range check must be performed.
+- Distrust any data that you extract from images or from external sources. E.g. if the metadata of an image gives you an offset of another information inside that file, do not assume that this offset will not result in an out off bounds read.
+- New code must not assume the endianes and the wordsize of the system it is being run on. I.e. don't assume that  `sizeof(int) = 8` or that the following will work:
+```cpp
+const uint32_t some_var = get_var();
+const uint16_t lower_2_bytes = (const uint16_t*) &some_var;
+```
+since this will give you the upper two bytes on big endian systems.
```
Comment (D4N) [inline CONTRIBUTING.md]: I think this will not be properly indented due to the code block. Maybe we should just remove the example, so that the text will look better.

File: CONTRIBUTING.md (hunk 1 above)
Comment (D4N) [reply CONTRIBUTING.md]: Yeah I know, github recommends that, however I think most of the stuff in there is covered by their terms of service anyway. Also, we can hope that people will be decent unless we're proven wrong.

**Comment by D4N** [comment]
I think we should also add that the PRs should be kept at a manageable size, that the code should be simple to read & understand, that documentation must be added and we should decide how commits should be made.

File: CONTRIBUTING.md (hunk 2 above)
Comment (piponazo) [reply CONTRIBUTING.md]: I am using [Haroopad](http://pad.haroopress.com/) for editing the markdown text and it looks good.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+4. Make your change. Add tests for your change. Make the tests pass.
+5. Create the Pull Request (PR).
+	a. Suggest a reviewer if you know that some of the mantainers know about the specific change.
+    b. Check if the CI checks pass.
+    c. If your PR lives for a long time, try to avoid to press the button "Update branch" in the Pull Request view. We prefer rebasing in top of master, instead of merging master into branches.
+
+## Pull Request guidelines
+- PRs should be kept at a manageable size. Try to focus in just one goal per PR. If you find yourself doing several things that in a PR that were not expected, try to split the different tasks in different PRs.
+- Commit messages should as informative and concise as possible.
```
Comment (D4N) [inline CONTRIBUTING.md]: i usually try to stick to the following rule:
The first line of the commit message should be < 80 characters and describe the commit briefly. If the 80 characters are too short for a summary, consider splitting the commit.
Below that, add one blank line and then a more detailed explanation if required.

This has the advantage, that `git --oneline --graph` shows a good overview and the commit messages are not wrapped on github, yet additional information is not lost.

Comment (tbeu) [reply CONTRIBUTING.md]: That is the default git recommendation, right.

**Comment by D4N** [comment]
I think this is more or less complete. I'd maybe add also that commits should always change one logical unit, so that cherry-picking & reverting is simple.

File: CONTRIBUTING.md (hunk 3 above)
Comment (piponazo) [reply CONTRIBUTING.md]: Right, I always try to do the same, but for some people this is not obvious. I will add it

**Comment by piponazo** [comment]
Latest suggestions added and I also squashed the commits into a unique one.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# How to contribute #
+
+All help is appreciated. All contributors are unpaid volunteers. If you really want to see something added to Exiv2 - volunteer!
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: I'd propose to skip this sentences about volunteering and payments. We never know...

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+## General workflow for contributing ##
+
+1. Fork, then clone the repo:
+```
+git clone git@github.com:your-username/exiv2.git
+```
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: That is standard for GitHub hosted repos and documented at https://guides.github.com/activities/forking/, for example.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+2. Configure the project and build (More details about how to run CMake at XXX) :
+```
+mkdir build && cd build
+cmake .. && cmake --build . --config Release
+```
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: That misses the autotools and MSVC build process as alternatives. Could also be an extra file.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+make tests			# Application tests
+./bin/unit_tests	# Unit tests
+```
+
+4. Make your change. Add tests for your change. Make the tests pass.
+5. Create the Pull Request (PR).
+	a. Suggest a reviewer if you know that some of the mantainers know about the specific change.
+    b. Check if the CI checks pass.
+    c. If your PR lives for a long time, try to avoid to press the button "Update branch" in the Pull Request view. We prefer rebasing in top of master, instead of merging master into branches.
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: Indents for `a.` do not match. Reviewers can only be suggested by project members, right?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+## Pull Request guidelines
+- PRs should be kept at a manageable size. Try to focus in just one goal per PR. If you find yourself doing several things that in a PR that were not expected, try to split the different tasks in different PRs.
+- Commits should always change one logical unit, so that cherry-picking & reverting is simple.
+- Commit messages should as informative and concise as possible. The first line of the commit message should be < 80 characters and describe the commit briefly. If the 80 characters are too short for a summary, consider splitting the commit. Below the short summary add one blank line and then a more detailed explanation if required.
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: OK for me.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+## Code guidelines ##
+### General ###
+- All new code must be properly tested: via unit tests (based on the Gtest framework; see `$REPO/unitTest`) or system tests (scripts exercising the main exiv2 application; see `$REPO/test`).
+- Code should be simple to read and to understand.
+- Do not invoke undefined behavior! Ensure that with UBSAN, i.e. compile your code with `-fsanitize=undefined` and run the test suite.
+- Ensure that your code has no memory errors. Use ASAN for that, i.e. compile your code with `-fsanitize=address`.
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: UBSAN/ASAN are strong requirements and exclude MSVC users. I'd mark it optionally,

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+### Relevant to the Exiv2 domain ###
+- All new code that is added must be resistant to integer overflows, thus if you multiply, add, subtract, divide or bitshift integers you must ensure that no overflow can occur. Please keep in mind that signed integer overflow is undefined behavior, thus you must check for overflows before performing the arithmetic operation, otherwise the compiler is free to optimize your check after the overflow away (this has happened already).
+- All new code must be resistant to buffer overflows. Thus before you access arrays a range check must be performed.
+- Distrust any data that you extract from images or from external sources. E.g. if the metadata of an image gives you an offset of another information inside that file, do not assume that this offset will not result in an out off bounds read.
+- New code must not assume the endianes and the wordsize of the system it is being run on. I.e. don't assume that  `sizeof(int) = 8` or that the following will work:
+```cpp
+const uint32_t some_var = get_var();
+const uint16_t lower_2_bytes = (const uint16_t*) &some_var;
+```
```
Comment (tbeu) [inline CONTRIBUTING.md]: As a site note: Detecting integer overflows in a C++ portable way is pain. I recently dicovered https://github.com/nemequ/portable-snippets/tree/master/safe-math which I intent to use in my other projects.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+```cpp
+// clang-format off
+    void    unformatted_code  ;
+// clang-format on
+```
+
+More information about clang:
+- Link to [clang-format](https://clang.llvm.org/docs/ClangFormat.html) tool.
+- Link to the [clang-format option](https://clang.llvm.org/docs/ClangFormatStyleOptions.html).
+
```
Comment (tbeu) [inline CONTRIBUTING.md]: Again, strong requirement. Code formatting still can be applied from a tproject member afterwards by pushing to the contributor's branch. I'd mark it optionally.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+## Other points to consider ##
+
+- [Old document](http://dev.exiv2.org/projects/exiv2/wiki/Contributing_to_Exiv2) with more information about contributions.
```
Comment (tbeu) [inline CONTRIBUTING.md]: I'd not link here, but rather copy the main contents from there, if relevant.

Comment (tbeu) [reply CONTRIBUTING.md]: Generally, a lot of instructions. I'd prefer such simple instructions as stated in https://github.com/google/honggfuzz/blob/master/CONTRIBUTING, for example.

**Review by tbeu — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 6 above)
Comment (piponazo) [reply CONTRIBUTING.md]: I can mention that this is the preferred/default/(just the cmake way) way to work in Exiv2 and point to other existing ways that are explained in the README.md. 

My idea is to keep this document as simple as possible, and focus primarily in how to contribute to the project. Details about all the possibilities to configure/build the project can be found in other text files.

File: CONTRIBUTING.md (hunk 5 above)
Comment (piponazo) [reply CONTRIBUTING.md]: I could add that that link at the beginning of the section, but I think it is interesting anyways to keep these points as a quick reference workflow to contribute.

File: CONTRIBUTING.md (hunk 4 above)
Comment (piponazo) [reply CONTRIBUTING.md]: I took this from the link you recommended me to read:
http://dev.exiv2.org/projects/exiv2/wiki/Contributing_to_Exiv2

I will remove the sentence and let you add whatever you want later.

File: CONTRIBUTING.md (hunk 7 above)
Comment (piponazo) [reply CONTRIBUTING.md]: Good point. I think you are right. I will remove that point and just mention that a project member will review the PR ASAP.

File: CONTRIBUTING.md (hunk 11 above)
Comment (piponazo) [reply CONTRIBUTING.md]: Note that we are not enforcing this:

```
One can apply this format to new or existing code by using the **clang-format** command-line tool, or by using some integration plugins provided by some editors or IDEs.
```
I will rephrase it to make it more clear.

File: CONTRIBUTING.md (hunk 12 above)
Comment (piponazo) [reply CONTRIBUTING.md]: Removing section

File: CONTRIBUTING.md (hunk 10 above)
Comment (D4N) [reply CONTRIBUTING.md]: I know, especially if you intend to support other compilers than gcc and clang which provide very efficient builtins for that. I intend to provide something portable and usable in the near future.

File: CONTRIBUTING.md (hunk 9 above)
Comment (D4N) [reply CONTRIBUTING.md]: The requirement that no memory errors and no undefined behavior occurs must stay. We can make the usage of ASAN & UBSAN a suggestion but I intend to add them to the CI.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -37,8 +34,8 @@ make tests			# Application tests
 ### General ###
 - All new code must be properly tested: via unit tests (based on the Gtest framework; see `$REPO/unitTest`) or system tests (scripts exercising the main exiv2 application; see `$REPO/test`).
 - Code should be simple to read and to understand.
-- Do not invoke undefined behavior! Ensure that with UBSAN, i.e. compile your code with `-fsanitize=undefined` and run the test suite.
-- Ensure that your code has no memory errors. Use ASAN for that, i.e. compile your code with `-fsanitize=address`.
+- [Optional] Try to avoid undefined behavior. Ensure that with UBSAN, i.e. compile your code with `-fsanitize=undefined` and run the test suite.
+- [Optional] Ensure that your code has no memory errors. Use ASAN for that, i.e. compile your code with `-fsanitize=address`.
```
Comment (D4N) [inline CONTRIBUTING.md]: No memory errors are not optional, we have enough security issues, we don't want new ones. ASAN can stay optional for MSVC users.

Comment (piponazo) [reply CONTRIBUTING.md]: changed

**Review by D4N — APPROVED** [review]
This is fine by me. If some information turn out to be missing, we can always add them later.
